File: source/pbat/sim/algorithm/vbd/Anderson.cpp
#include "Anderson.h"

namespace pbat::sim::algorithm::vbd {

void AndersonParams::Serialize(io::Archive& archive, bool bMinimal) const
{
    io::Archive group = archive["pbat.sim.algorithm.vbd.AndersonParams"];
    group.WriteMetaData("m", m);
    group.WriteMetaData("beta", beta);
    group.WriteMetaData("codNumericalZero", codNumericalZero);
    if (not bMinimal)
    {
        group.WriteData("Fk", Fk);
        group.WriteData("Xk", Xk);
        group.WriteData("xkm1", xkm1);
        group.WriteData("fk", fk);
        group.WriteData("fkm1", fkm1);
        group.WriteData("gammak", gammak);
    }
}

void AndersonParams::Deserialize(io::Archive const& archive)
{
    io::Archive group = archive["pbat.sim.algorithm.vbd.AndersonParams"];
    if (group.HasMetaData("m"))
        m = group.ReadMetaData<Index>("m");
    if (group.HasMetaData("beta"))
        beta = group.ReadMetaData<Scalar>("beta");
    if (group.HasMetaData("codNumericalZero"))
        codNumericalZero = group.ReadMetaData<Scalar>("codNumericalZero");
    if (group.HasData("Fk"))
        Fk = group.ReadData<MatrixX>("Fk");
    if (group.HasData("Xk"))
        Xk = group.ReadData<MatrixX>("Xk");
    if (group.HasData("xkm1"))
        xkm1 = group.ReadData<VectorX>("xkm1");
    if (group.HasData("fk"))
        fk = group.ReadData<VectorX>("fk");
    if (group.HasData("fkm1"))
        fkm1 = group.ReadData<VectorX>("fkm1");
    if (group.HasData("gammak"))
        gammak = group.ReadData<VectorX>("gammak");
}

void AndersonParams::AllocateIfNeeded(Index n)
{
    Fk.resize(n, m);
    Xk.resize(n, m);
    xkm1.resize(n);
    fk.resize(n);
    fkm1.resize(n);
    gammak.resize(m);
}

} // namespace pbat::sim::algorithm::vbd

#include "pbat/graph/Adjacency.h"
#include "pbat/graph/Color.h"
#include "pbat/graph/Mesh.h"
#include "pbat/physics/StableNeoHookeanEnergy.h"

#include <doctest/doctest.h>

namespace {

struct AndersonTestSetup
{
    using ElasticEnergyType = pbat::physics::StableNeoHookeanEnergy<3>;
    using FemElastoDynamics = pbat::sim::algorithm::common::FemElastoDynamics<ElasticEnergyType>;
    using ScalarType        = typename FemElastoDynamics::ScalarType;
    using IndexType         = typename FemElastoDynamics::IndexType;

    FemElastoDynamics dynamics;
    pbat::sim::algorithm::vbd::Params vbdParams;
    pbat::sim::algorithm::vbd::AndersonParams andersonParams;
    pbat::sim::contact::MeshDynamics<ScalarType, IndexType> meshDynamics;
    pbat::MatrixX X;
    pbat::geometry::Device device;
};

AndersonTestSetup SetupAndersonTest(pbat::Index maxIters = 10)
{
    using namespace pbat;
    AndersonTestSetup setup;

    // Cube mesh
    setup.X.resize(3, 8);
    IndexMatrixX C(4, 5);
    // clang-format off
    setup.X << 0., 1., 0., 1., 0., 1., 0., 1.,
               0., 0., 1., 1., 0., 0., 1., 1.,
               0., 0., 0., 0., 1., 1., 1., 1.;
    C << 0, 3, 5, 6, 0,
         1, 2, 4, 7, 5,
         3, 0, 6, 5, 3,
         5, 6, 0, 3, 6;
    // clang-format on

    // Construct FEM dynamics
    setup.dynamics.Construct(setup.X, C);

    // Adjacency structures
    sim::algorithm::vbd::VertexElementAdjacencyGraph(
        C,
        setup.X.cols(),
        setup.vbdParams.GVGp,
        setup.vbdParams.GVGe,
        setup.vbdParams.GVGilocal);

    // Vertex colors
    auto eOrdering  = graph::EGreedyColorOrderingStrategy::LargestDegree;
    auto eSelection = graph::EGreedyColorSelectionStrategy::LeastUsed;
    sim::algorithm::vbd::VertexColors(
        C,
        setup.X.cols(),
        graph::EGreedyColorOrderingStrategy::LargestDegree,
        graph::EGreedyColorSelectionStrategy::LeastUsed,
        setup.vbdParams.GVVp,
        setup.vbdParams.GVVadj,
        setup.vbdParams.colors);

    // VBD params
    setup.vbdParams
        .WithVertexColors(setup.vbdParams.GVVp, setup.vbdParams.GVVadj, setup.vbdParams.colors)
        .WithMaximumIterations(maxIters)
        .WithHessianSingularUnder(Scalar{1e-6})
        .Construct();

    // Anderson params
    setup.andersonParams.m                = 5;
    setup.andersonParams.beta             = Scalar(1);
    setup.andersonParams.codNumericalZero = Scalar(1e-10);

    // Mesh contact dynamics (minimal setup for testing)
    IndexVectorX XCC(setup.X.cols()), ECC(C.cols()), Xord(setup.X.cols()), Eord(C.cols());
    Eigen::Index const nComponents =
        graph::SortedConnectedComponentOrdering(setup.X, C, XCC, ECC, Xord, Eord);
    graph::ReindexMeshByConnectedComponents(setup.X, C, XCC, ECC, Xord, Eord);
    sim::contact::MultiMesh<Index> multiMesh(C.bottomRows<4>(), XCC, nComponents);
    setup.meshDynamics.SetDynamicGeometry(setup.X, std::move(multiMesh));
    setup.meshDynamics.Initialize(setup.device);
    return setup;
}

} // namespace

TEST_CASE("[sim][algorithm][vbd] Anderson")
{
    using namespace pbat;
    // Arrange
    auto setup = SetupAndersonTest(10);
    // Act
    setup.dynamics.SetInitialConditions(setup.dynamics.x, setup.dynamics.v);
    setup.dynamics.SetupTimeIntegrationOptimization();
    Scalar f0  = setup.dynamics.Objective(setup.dynamics.x);
    VectorX g0 = setup.dynamics.Gradient(setup.dynamics.x);
    sim::algorithm::vbd::InitializeSolve(
        setup.dynamics,
        setup.meshDynamics,
        setup.vbdParams,
        setup.andersonParams);
    sim::algorithm::vbd::Solve(
        setup.dynamics,
        setup.meshDynamics,
        setup.vbdParams,
        setup.andersonParams);
    // Assert
    auto constexpr zero = Scalar{1e-4};
    auto xt             = setup.dynamics.bdf.CurrentState(0).reshaped(
        setup.dynamics.x.rows(),
        setup.dynamics.x.cols());
    MatrixX dx                           = setup.dynamics.x - xt;
    bool const bVerticesFallUnderGravity = (dx.row(2).array() < Scalar{0}).all();
    CHECK(bVerticesFallUnderGravity);
    bool const bVerticesOnlyFall = (dx.topRows(2).array().abs() < zero).all();
    CHECK(bVerticesOnlyFall);
    Scalar f = setup.dynamics.Objective(setup.dynamics.x);
    CHECK_LT(f, f0);
    VectorX g     = setup.dynamics.Gradient(setup.dynamics.x);
    Scalar g0norm = g0.norm();
    Scalar gnorm  = g.norm();
    CHECK_LT(gnorm, g0norm);
}

#include <tbb/global_control.h>

TEST_CASE("[sim][algorithm][anderson] Anderson Sandbox")
{
    // using namespace pbat;
    // auto archive            = io::Archive("sandbox.h5", HighFive::File::AccessMode::ReadOnly);
    // using ElasticEnergyType = pbat::physics::StableNeoHookeanEnergy<3>;
    // sim::algorithm::common::FemElastoDynamics<ElasticEnergyType> fem;
    // fem.Deserialize(archive["fem"]);
    // sim::contact::MeshDynamics<Scalar, Index> contact;
    // contact.Deserialize(archive["contact"]);
    // sim::algorithm::vbd::Params params;
    // params.Deserialize(archive["vbd/params"]);
    // sim::algorithm::vbd::AndersonParams andersonParams;
    // andersonParams.Deserialize(archive["vbd/anderson_params"]);
    // geometry::Device device{geometry::DeviceConfig{}.WithVerbosity(4)};
    // contact.Initialize(device);
    // tbb::global_control gc(tbb::global_control::max_allowed_parallelism, 1);
    // for (auto t = 0; t < 50; ++t)
    // {
    //     fem.SetupTimeIntegrationOptimization(
    //         sim::dynamics::EFemElastoDynamicsTimeStepInitialization::TrajectoryWithExternalLoad);
    //     sim::algorithm::vbd::InitializeSolve(fem, contact, params, andersonParams);
    //     sim::algorithm::vbd::Solve(fem, contact, params, andersonParams);
    //     fem.Step();
    // }
}

File: python/tools/vbd/ui/solvers/anderson.py
# type: ignore
from pbatoolkit import pbat
from ..params import ParameterObject
import polyscope as ps
import polyscope.imgui as imgui
from .base import BaseSolver
import typing
import numpy as np


class Params:
    _vbd_params: pbat.sim.algorithm.vbd.Params
    _anderson_params: pbat.sim.algorithm.vbd.AndersonParams

    def __init__(self):
        self._vbd_params = pbat.sim.algorithm.vbd.Params()
        self._anderson_params = pbat.sim.algorithm.vbd.AndersonParams()

    @property
    def vbd_params(self):
        return self._vbd_params

    @vbd_params.setter
    def vbd_params(self, value: pbat.sim.algorithm.vbd.Params):
        self._vbd_params = value

    @property
    def anderson_params(self):
        return self._anderson_params

    @anderson_params.setter
    def anderson_params(self, value: pbat.sim.algorithm.vbd.AndersonParams):
        self._anderson_params = value


class AndersonSolver(BaseSolver):
    _params: ParameterObject

    def __init__(self):
        super().__init__("Anderson")
        self._params = ParameterObject(
            Params(), {"vbd_params": None, "anderson_params": None}
        )

    def draw(self):
        imgui.PushID(self._name)
        if imgui.TreeNode(f"{type(self._params.params).__name__}"):
            self._params.draw()
            imgui.TreePop()
        imgui.PopID()

    def on_simulation_scenario_created(
        self,
        fem: pbat.sim.dynamics.FemElastoDynamics,
        contact: pbat.sim.contact.MeshDynamics,
    ):
        params: pbat.sim.algorithm.vbd.Params = self._params.params.vbd_params
        n_nodes = fem.X.shape[1]
        GVGp, GVGe, GVGilocal = pbat.sim.algorithm.vbd.vertex_element_adjacency_graph(
            fem.E, n_nodes
        )
        GVVp, GVVadj, colors = pbat.sim.algorithm.vbd.vertex_colors(fem.E, n_nodes)
        params.with_vertex_element_adjacency_graph(
            GVGp, GVGe, GVGilocal
        ).with_vertex_colors(GVVp, GVVadj, colors).construct()

    def solve(
        self,
        fem: pbat.sim.dynamics.FemElastoDynamics,
        contact: pbat.sim.contact.MeshDynamics,
        callback: typing.Callable[None, None] | None = None,
    ):
        params: Params = self._params.params
        vbd = params.vbd_params
        anderson = params.anderson_params
        if callback is None:
            callback = lambda: None
        callback()
        # Update Anderson solver for most recent SAL contact framework
        pbat.sim.algorithm.vbd.initialize_solve(fem, contact, vbd, anderson)
        callback()
        for k in range(vbd.n_max_iters):
            callback()
            pbat.sim.algorithm.vbd.linearize_constraints(fem, contact)
            converged = pbat.sim.algorithm.vbd.check_convergence(fem, contact, vbd)
            if converged:
                break
            pbat.sim.algorithm.vbd.prepare_subproblem(fem, contact, vbd)
            for kp in range(vbd.n_subproblem_max_iters):
                pbat.sim.algorithm.vbd.iterate(fem, contact, vbd, anderson)
            pbat.sim.algorithm.vbd.finalize_subproblem(fem, contact, vbd)
        callback()
        fem.back_substitute_integrated_positions_into_velocities()

    def serialize(self, archive: pbat.io.Archive):
        params: Params = self._params.params
        vbd: pbat.sim.algorithm.vbd.Params = params.vbd_params
        anderson: pbat.sim.algorithm.vbd.AndersonParams = params.anderson_params
        vbd.serialize(archive, minimal=True)
        anderson.serialize(archive, minimal=True)

    def deserialize(self, archive: pbat.io.Archive):
        params: Params = self._params.params
        vbd: pbat.sim.algorithm.vbd.Params = params.vbd_params
        anderson: pbat.sim.algorithm.vbd.AndersonParams = params.anderson_params
        vbd.deserialize(archive)
        anderson.deserialize(archive)

    def serialize_problem(
        self,
        archive: pbat.io.Archive,
        fem: pbat.sim.dynamics.FemElastoDynamics,
        contact: pbat.sim.contact.MeshDynamics,
    ):
        params: Params = self._params.params
        vbd: pbat.sim.algorithm.vbd.Params = params.vbd_params
        anderson: pbat.sim.algorithm.vbd.AndersonParams = params.anderson_params
        fem.serialize(archive["fem"])
        contact.serialize(archive["contact"])
        vbd.serialize(archive["vbd/params"], minimal=False)
        anderson.serialize(archive["vbd/anderson_params"], minimal=False)
